# Exercise 1

This exercise is based on the content in `GSP538/python_refesh/py_refresher.ipynb`

NOTE: To make doing the assignment easier:
- Open this notebook and the refresher notebook and display them side-by-side.
  - If they are both currently open in separate tabs, you can drag one tab until the split screen is highlighted within the VS Code
  - You can also right-click one file in the Explorer pane and choose "Open to the Side"
- Collapse the Explorer pane.

#### Answer the following written questions

There is a blank Markdown cell after each question for your answer (double click in the blank cell to type your answer). Be sure to run your Markdown cells to format your answers.

1. Provide at least two differences between the NumPy and Pandas packages.

1. **NumPy** (Numerical Python) provides multidimensional arrays and efficient numerical operations. **Pandas** (Panel Data) is built on top of NumPy and adds user-friendly features such as named rows and columns, missing data management, and the ability to read/write files in various formats. 
2. NumPy's primary data structure is the `Array`, while Pandas uses `DataFrame` and `Series`. Pandas DataFrames are essentially NumPy Arrays with additional capabilities like grouping, merging, and joining.

2. How is "dot notation" is used in Python?

Dot notation uses the period (`.`) to chain an object to its attributes or methods. The syntax is `object.attribute` or `object.method()`. It is used for three main purposes in Python: (1) accessing methods or attributes of an object (e.g., `s.upper()`), (2) accessing contents of a module or package (e.g., `math.sqrt(16)`), and (3) accessing columns or methods of a DataFrame (e.g., `df.population`).

3. What do the `dir()` and `type()` functions do in Python?

`type()` returns the type (class) of an object. For example, `type(2)` returns `int`, `type('dog')` returns `str`, and `type(df)` returns `pandas.core.frame.DataFrame`. `dir()` lists all the attributes and methods available on an object. For example, `dir(s)` on a string shows all string methods like `upper`, `lower`, etc. Called without arguments, `dir()` shows all objects defined in the current session.

4. What does "namespace" mean in Python and why is it important to keep different packages in their own namespace?

A namespace is a collection of currently defined names along with information about the object each name references. It is important to keep different packages in their own namespaces to prevent name collisions. For example, both `math` and `numpy` have a `sqrt` function — by keeping them in separate namespaces (`math.sqrt` vs `np.sqrt`), we can use both without conflict. This is why `from module import *` is discouraged, as it dumps all names into the main namespace and can silently overwrite existing functions.

5. What does it mean when a programming language is said to be "zero-indexed"? Is Python zero-indexed?

A zero-indexed language means that the first element of a sequence is at position 0, the second element is at position 1, and so on. Yes, Python is zero-indexed. For example, given `s = 'hello'`, `s[0]` returns `'h'` (the first character) and `s[1]` returns `'e'` (the second character).

6. Once a function is defined, you can then run it. When running a function (not defining a function), why can you sometimes skip the argument name and sometimes you cannot?

You can skip the argument name when passing arguments in the correct positional order defined by the function signature. If you pass arguments out of order, you must use the argument names (keyword arguments). Also, once you start naming arguments, all subsequent arguments must also be named. For example, with `round(number, ndigits)`, you can call `round(10.867, 2)` positionally, or `round(ndigits=2, number=10.867)` using keyword arguments in any order.

7. If you have a Pandas DataFrame named `my_data` that has a column named `population`, what is the Python syntax for two ways you can access the column?

Two ways to access the `population` column:
1. **Square bracket notation:** `my_data['population']`
2. **Dot notation:** `my_data.population`

8. Read the "Generative Artificial Intelligence" section of the course syllabus. For context, this question and the ones above are "narrative based questions", the questions below are "code based questions." How are code questions scored? How can AI be used for narrative questions?

Code-based questions are scored based on whether the code runs correctly and produces the expected output. For narrative-based questions, AI can be used as a learning aid to help understand concepts, but answers should be written in the student's own words to demonstrate understanding of the material.

#### The following questions require you to run Python code.

For these questions, you will be using an extract from the GeoNames database ([https://www.geonames.org](https://www.geonames.org)). The raw data from GeoNames has been lightly processed so that you will be working with all cities in the US with population of at least 15,000 people.

Run the following cell to import the data.

In [13]:
import pandas as pd  #L1

data = pd.read_csv('data/us_cities15000.txt', sep='\t')    #L2

9. The cell above contains two lines of code. Give a detailed explanation of each line. Hint: `\t` means tab (like the tab key on the keyboard); you might find [this page](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html) helpful.

**L1:** `import pandas as pd` imports the Pandas library and gives it the alias `pd`. This means we can access all Pandas functions using `pd.function_name()` instead of typing out `pandas.function_name()` each time.

**L2:** `data = pd.read_csv('data/us_cities15000.txt', sep='\t')` uses the Pandas `read_csv()` function to read a tab-separated text file located at `data/us_cities15000.txt`. The `sep='\t'` argument specifies that columns are separated by tab characters (rather than the default comma). The resulting DataFrame is stored in a variable called `data`.

10. Object type
    - Use Python code to determine what type of object `data` is
    - State what the Python output means in words

In [14]:
type(data)

pandas.DataFrame

The output `pandas.core.frame.DataFrame` means that the variable `data` is a Pandas DataFrame object — a two-dimensional tabular data structure with labeled rows and columns.

11. Rows and columns.
    - Use Python code to determine how many rows and columns are in `data`
    - State what the output means in words

In [15]:
data.shape

(3269, 19)

The output `(3269, 19)` means that the `data` DataFrame has 3,269 rows and 19 columns. Each row represents a US city with a population of at least 15,000 people, and each column represents a different attribute of that city.

12. Column names.
    - Use Python code to get the column names of `data`. Note: only return the column names, not the whole dataframe.
    - How many names are listed in the output? Does this match the result from the previous question?

In [16]:
data.columns

Index(['geonameid', 'name', 'asciiname', 'alt_names', 'latitude', 'longitude',
       'feature_class', 'feature_code', 'country_code', 'cc2', 'admin1_code',
       'admin2_code', 'admin3_code', 'admin4_code', 'population', 'elevation',
       'dem', 'timezone', 'mod_date'],
      dtype='str')

There are 19 column names listed in the output. Yes, this matches the result from the previous question, which showed 19 columns in the `data.shape` output.

13. Run the following cell to see the first five rows of `data`. Notice that you can scroll right and left to see all the data in the output.

    - Now modify the following cell to show the first 10 rows of `data`. This was not shown in the notebook you reviewed before starting this exercise. Go to Google and search the following to get help on how to do this: `pandas dataframe head`. Or use the AI assistant in VS Code to read the documentation.
    - Explain in your own words why the original cell gave five rows.

In [17]:
data.head()

,geonameid,name,asciiname,alt_names,latitude,longitude,feature_class,feature_code,country_code,cc2,admin1_code,admin2_code,admin3_code,admin4_code,population,elevation,dem,timezone,mod_date
0,4046704,Fort Hunt,Fort Hunt,NaN,38.73289,-77.05803,P,PPL,US,NaN,VA,59.0,NaN,NaN,16045,10.0,16,America/New_York,2011-05-14
1,4048023,Bessemer,Bessemer,"Besemer,Bessemer,bei se mo,bes'semara,beseomeo...",33.40178,-86.95444,P,PPL,US,NaN,AL,73.0,NaN,NaN,26730,155.0,161,America/Chicago,2017-03-09
2,4048662,Paducah,Paducah,"PAH,Padaka,Padjuka,Paducah,Paduka,Pekin,badwka...",37.08339,-88.60005,P,PPLA2,US,NaN,KY,145.0,NaN,NaN,24864,104.0,104,America/Chicago,2018-01-03
3,4049979,Birmingham,Birmingham,"BHM,Bermincham,Bermingkham,Birmingam,Birmingam...",33.52066,-86.80249,P,PPLA2,US,NaN,AL,73.0,NaN,NaN,212461,187.0,187,America/Chicago,2022-11-19
4,4054378,Center Point,Center Point,"Senter Pojnt,Sentr-Pojnt,sentara po'inta,sntr ...",33.64566,-86.68360,P,PPL,US,NaN,AL,73.0,NaN,NaN,16655,253.0,256,America/Chicago,2017-03-09


In [18]:
data.head(10)

,geonameid,name,asciiname,alt_names,latitude,longitude,feature_class,feature_code,country_code,cc2,admin1_code,admin2_code,admin3_code,admin4_code,population,elevation,dem,timezone,mod_date
0,4046704,Fort Hunt,Fort Hunt,NaN,38.73289,-77.05803,P,PPL,US,NaN,VA,59.0,NaN,NaN,16045,10.0,16,America/New_York,2011-05-14
1,4048023,Bessemer,Bessemer,"Besemer,Bessemer,bei se mo,bes'semara,beseomeo...",33.40178,-86.95444,P,PPL,US,NaN,AL,73.0,NaN,NaN,26730,155.0,161,America/Chicago,2017-03-09
2,4048662,Paducah,Paducah,"PAH,Padaka,Padjuka,Paducah,Paduka,Pekin,badwka...",37.08339,-88.60005,P,PPLA2,US,NaN,KY,145.0,NaN,NaN,24864,104.0,104,America/Chicago,2018-01-03
3,4049979,Birmingham,Birmingham,"BHM,Bermincham,Bermingkham,Birmingam,Birmingam...",33.52066,-86.80249,P,PPLA2,US,NaN,AL,73.0,NaN,NaN,212461,187.0,187,America/Chicago,2022-11-19
4,4054378,Center Point,Center Point,"Senter Pojnt,Sentr-Pojnt,sentara po'inta,sntr ...",33.64566,-86.68360,P,PPL,US,NaN,AL,73.0,NaN,NaN,16655,253.0,256,America/Chicago,2017-03-09
5,4057835,Cullman,Cullman,"Kallmen,Kalman,Kalmen,Kulman,ka er man,kalamya...",34.17482,-86.84361,P,PPLA2,US,NaN,AL,43.0,NaN,NaN,15350,242.0,244,America/Chicago,2022-11-19
6,4058219,Daphne,Daphne,"Dafni,Daphne,Daphney,Hollywood,da fu ni,dfny ...",30.60353,-87.90360,P,PPL,US,NaN,AL,3.0,NaN,NaN,24896,49.0,56,America/Chicago,2019-09-24
7,4058553,Decatur,Decatur,"DCU,Decatur,Dekejter,Dekejtur,decatur,dekatara...",34.60593,-86.98334,P,PPLA2,US,NaN,AL,103.0,NaN,NaN,55437,171.0,175,America/Chicago,2022-11-19
8,4059102,Dothan,Dothan,"Cawthons Cowpen,DHN,Dotan,Dothan,Dothen,Poplar...",31.22323,-85.39049,P,PPLA2,US,NaN,AL,69.0,NaN,NaN,68567,98.0,102,America/Chicago,2022-11-19
9,4059870,East Florence,East Florence,"East Florence,Sweetwater",34.80953,-87.64947,P,PPL,US,NaN,AL,77.0,NaN,NaN,35733,167.0,171,America/Chicago,2006-01-17


The original `data.head()` call gave five rows because the default value of the `n` parameter in the `head()` method is 5. When no argument is provided, `head()` returns the first 5 rows of the DataFrame.

14. Compute a new column.
    - Add a column to `data` named `elevation_ft` containing the elevation of each city in feet. Hint: the `elevation` column is in meters.
    - After answering making the computation, add the line `data.head()` to the cell to display the result.
    - Briefly explain how you solved this question.

In [19]:
data['elevation_ft'] = data['elevation'] * 3.28084
data.head()

,geonameid,name,asciiname,alt_names,latitude,longitude,feature_class,feature_code,country_code,cc2,admin1_code,admin2_code,admin3_code,admin4_code,population,elevation,dem,timezone,mod_date,elevation_ft
0,4046704,Fort Hunt,Fort Hunt,NaN,38.73289,-77.05803,P,PPL,US,NaN,VA,59.0,NaN,NaN,16045,10.0,16,America/New_York,2011-05-14,32.80840
1,4048023,Bessemer,Bessemer,"Besemer,Bessemer,bei se mo,bes'semara,beseomeo...",33.40178,-86.95444,P,PPL,US,NaN,AL,73.0,NaN,NaN,26730,155.0,161,America/Chicago,2017-03-09,508.53020
2,4048662,Paducah,Paducah,"PAH,Padaka,Padjuka,Paducah,Paduka,Pekin,badwka...",37.08339,-88.60005,P,PPLA2,US,NaN,KY,145.0,NaN,NaN,24864,104.0,104,America/Chicago,2018-01-03,341.20736
3,4049979,Birmingham,Birmingham,"BHM,Bermincham,Bermingkham,Birmingam,Birmingam...",33.52066,-86.80249,P,PPLA2,US,NaN,AL,73.0,NaN,NaN,212461,187.0,187,America/Chicago,2022-11-19,613.51708
4,4054378,Center Point,Center Point,"Senter Pojnt,Sentr-Pojnt,sentara po'inta,sntr ...",33.64566,-86.68360,P,PPL,US,NaN,AL,73.0,NaN,NaN,16655,253.0,256,America/Chicago,2017-03-09,830.05252


I created a new column called `elevation_ft` by multiplying the existing `elevation` column (which is in meters) by 3.28084 (the number of feet per meter). Using Pandas, this operation is applied to every row at once: `data['elevation_ft'] = data['elevation'] * 3.28084`.

15. Create a new DataFrame named `data2` that contains only the following columns:
    - `geonameid`
    - `name`
    - `latitude`
    - `longitude`
    - `feature_code`
    - `admin1_code`
    - `population`
    - `elevation_ft`
    
    No written description needed for this question

In [20]:
data2 = data[['geonameid', 'name', 'latitude', 'longitude', 'feature_code', 'admin1_code', 'population', 'elevation_ft']]

16. Using `data2`, create a new DataFrame called `az` that contains just the Arizona cities.
    - After answering the question, add the line `az.head()` to the cell to display the result.
    - Hint: use `data2.head()` to figure out which column contains the state identifiers and how Arizona is likely defined in this dataset.
    - No written description needed for this question

In [21]:
az = data2.loc[data2['admin1_code'] == 'AZ']
az.head()

,geonameid,name,latitude,longitude,feature_code,admin1_code,population,elevation_ft
2365,5287262,Buckeye,33.37032,-112.58378,PPL,AZ,50876,869.42260
2366,5287565,Bullhead City,35.14778,-114.56830,PPL,AZ,39445,557.74280
2367,5288636,Casa Grande,32.87950,-111.75735,PPL,AZ,51460,1397.63784
2368,5288661,Casas Adobes,32.32341,-110.99510,PPL,AZ,66795,2411.41740
2369,5288786,Catalina Foothills,32.29785,-110.91870,PPL,AZ,50796,2575.45940


17. Arizona cities.
    - Using Python code, how many Arizona cites are in this dataset? Note: your code needs to return a single integer.
    - State your result in a sentence. 

In [22]:
az.shape[0]

56

There are 56 Arizona cities in this dataset with a population of at least 15,000 people.

18. Final result.
    - Extract all the cities from `az` with an elevation above 5000 feet. (Note: don't create a new DataFrame for this question, just return the appropriate rows.)
    - State what the output means in words

In [23]:
az.loc[az['elevation_ft'] > 5000]

,geonameid,name,latitude,longitude,feature_code,admin1_code,population,elevation_ft
2376,5294810,Flagstaff,35.19807,-111.65127,PPLA2,AZ,70320,6902.88736
2396,5309842,Prescott,34.54002,-112.46850,PPLA2,AZ,41899,5374.01592
2397,5309858,Prescott Valley,34.61002,-112.31572,PPL,AZ,42197,5026.24688


The output shows the 3 Arizona cities with an elevation above 5,000 feet: Flagstaff (approximately 6,903 ft), Prescott (approximately 5,374 ft), and Prescott Valley (approximately 5,026 ft). These are all cities in the northern/central highlands of Arizona.

Optional (just for fun). Using the `data` DataFrame, get the exact same answer as question 18 (same rows and columns) using just one line of code.

In [24]:
data.loc[(data['admin1_code'] == 'AZ') & (data['elevation_ft'] > 5000), ['geonameid', 'name', 'latitude', 'longitude', 'feature_code', 'admin1_code', 'population', 'elevation_ft']]

,geonameid,name,latitude,longitude,feature_code,admin1_code,population,elevation_ft
2376,5294810,Flagstaff,35.19807,-111.65127,PPLA2,AZ,70320,6902.88736
2396,5309842,Prescott,34.54002,-112.46850,PPLA2,AZ,41899,5374.01592
2397,5309858,Prescott Valley,34.61002,-112.31572,PPL,AZ,42197,5026.24688
